File: tests_deps/src/test/java/com/github/klee0kai/stone/test/lists/cachecontrol/gc/BumperGcTests.java
package com.github.klee0kai.stone.test.lists.cachecontrol.gc;

import com.github.klee0kai.stone.Stone;
import com.github.klee0kai.stone.closed.types.ListUtils;
import com.github.klee0kai.test.car.di.cachecontrol.gc.CarGcComponent;
import com.github.klee0kai.test.car.model.Bumper;
import org.junit.jupiter.api.Test;

import java.lang.ref.WeakReference;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class BumperGcTests {

    @Test
    public void createWorkCorrect() {
        // Given
        CarGcComponent DI = Stone.createComponent(CarGcComponent.class);

        //When
        List<WeakReference<Bumper>> bumperFactory = ListUtils.format(DI.bumpersModule().bumperFactory(), WeakReference::new);
        List<WeakReference<Bumper>> bumperWeak = ListUtils.format(DI.bumpersModule().bumperWeak(), WeakReference::new);
        List<WeakReference<Bumper>> bumperSoft = ListUtils.format(DI.bumpersModule().bumperSoft(), WeakReference::new);
        List<WeakReference<Bumper>> bumperStrong = ListUtils.format(DI.bumpersModule().bumperStrong(), WeakReference::new);

        //Then
        assertEquals(3, nonNullCount(bumperFactory));
        assertEquals(3, nonNullCount(bumperWeak));
        assertEquals(3, nonNullCount(bumperSoft));
        assertEquals(3, nonNullCount(bumperStrong));
    }


    @Test
    public void gcAllTest() {
        // Given
        CarGcComponent DI = Stone.createComponent(CarGcComponent.class);
        List<WeakReference<Bumper>> bumperFactory = ListUtils.format(DI.bumpersModule().bumperFactory(), WeakReference::new);
        List<WeakReference<Bumper>> bumperWeak = ListUtils.format(DI.bumpersModule().bumperWeak(), WeakReference::new);
        List<WeakReference<Bumper>> bumperSoft = ListUtils.format(DI.bumpersModule().bumperSoft(), WeakReference::new);
        List<WeakReference<Bumper>> bumperStrong = ListUtils.format(DI.bumpersModule().bumperStrong(), WeakReference::new);

        //When
        DI.gcAll();

        // Then
        assertEquals(0, nonNullCount(bumperFactory));
        assertEquals(0, nonNullCount(bumperWeak));
        assertEquals(0, nonNullCount(bumperSoft));
        assertEquals(0, nonNullCount(bumperStrong));
    }


    @Test
    public void gcWeakTest() {
        // Given
        CarGcComponent DI = Stone.createComponent(CarGcComponent.class);
        List<WeakReference<Bumper>> bumperFactory = ListUtils.format(DI.bumpersModule().bumperFactory(), WeakReference::new);
        List<WeakReference<Bumper>> bumperWeak = ListUtils.format(DI.bumpersModule().bumperWeak(), WeakReference::new);
        List<WeakReference<Bumper>> bumperSoft = ListUtils.format(DI.bumpersModule().bumperSoft(), WeakReference::new);
        List<WeakReference<Bumper>> bumperStrong = ListUtils.format(DI.bumpersModule().bumperStrong(), WeakReference::new);

        //When
        DI.gcWeak();

        // Then
        assertEquals(0, nonNullCount(bumperFactory));
        assertEquals(0, nonNullCount(bumperWeak));
        assertEquals(3, nonNullCount(bumperSoft));
        assertEquals(3, nonNullCount(bumperStrong));
    }

    @Test
    public void gcSoftTest() {
        // Given
        CarGcComponent DI = Stone.createComponent(CarGcComponent.class);
        List<WeakReference<Bumper>> bumperFactory = ListUtils.format(DI.bumpersModule().bumperFactory(), WeakReference::new);
        List<WeakReference<Bumper>> bumperWeak = ListUtils.format(DI.bumpersModule().bumperWeak(), WeakReference::new);
        List<WeakReference<Bumper>> bumperSoft = ListUtils.format(DI.bumpersModule().bumperSoft(), WeakReference::new);
        List<WeakReference<Bumper>> bumperStrong = ListUtils.format(DI.bumpersModule().bumperStrong(), WeakReference::new);

        //When
        DI.gcSoft();

        // Then
        assertEquals(0, nonNullCount(bumperFactory));
        assertEquals(0, nonNullCount(bumperWeak));
        assertEquals(0, nonNullCount(bumperSoft));
        assertEquals(3, nonNullCount(bumperStrong));
    }


    @Test
    public void gcStrongTest() {
        // Given
        CarGcComponent DI = Stone.createComponent(CarGcComponent.class);
        List<WeakReference<Bumper>> bumperFactory = ListUtils.format(DI.bumpersModule().bumperFactory(), WeakReference::new);
        List<WeakReference<Bumper>> bumperWeak = ListUtils.format(DI.bumpersModule().bumperWeak(), WeakReference::new);
        List<WeakReference<Bumper>> bumperSoft = ListUtils.format(DI.bumpersModule().bumperSoft(), WeakReference::new);
        List<WeakReference<Bumper>> bumperStrong = ListUtils.format(DI.bumpersModule().bumperStrong(), WeakReference::new);

        //When
        DI.gcStrong();

        // Then
        assertEquals(0, nonNullCount(bumperFactory));
        assertEquals(0, nonNullCount(bumperWeak));
        assertEquals(3, nonNullCount(bumperSoft));
        assertEquals(0, nonNullCount(bumperStrong));
    }

    @Test
    public void gcBumpers() {
        // Given
        CarGcComponent DI = Stone.createComponent(CarGcComponent.class);
        List<WeakReference<Bumper>> bumperFactory = ListUtils.format(DI.bumpersModule().bumperFactory(), WeakReference::new);
        List<WeakReference<Bumper>> bumperWeak = ListUtils.format(DI.bumpersModule().bumperWeak(), WeakReference::new);
        List<WeakReference<Bumper>> bumperSoft = ListUtils.format(DI.bumpersModule().bumperSoft(), WeakReference::new);
        List<WeakReference<Bumper>> bumperStrong = ListUtils.format(DI.bumpersModule().bumperStrong(), WeakReference::new);

        //When
        DI.gcBumpers();

        // Then
        assertEquals(0, nonNullCount(bumperFactory));
        assertEquals(0, nonNullCount(bumperWeak));
        assertEquals(0, nonNullCount(bumperSoft));
        assertEquals(0, nonNullCount(bumperStrong));
    }

    @Test
    public void gcRedBumpers() {
        // Given
        CarGcComponent DI = Stone.createComponent(CarGcComponent.class);
        List<WeakReference<Bumper>> bumperFactory = ListUtils.format(DI.bumpersModule().bumperFactory(), WeakReference::new);
        List<WeakReference<Bumper>> bumperWeak = ListUtils.format(DI.bumpersModule().bumperWeak(), WeakReference::new);
        List<WeakReference<Bumper>> bumperSoft = ListUtils.format(DI.bumpersModule().bumperSoft(), WeakReference::new);
        List<WeakReference<Bumper>> bumperStrong = ListUtils.format(DI.bumpersModule().bumperStrong(), WeakReference::new);

        //When
        DI.gcRedBumpers();

        // Then
        assertEquals(0, nonNullCount(bumperFactory));
        assertEquals(0, nonNullCount(bumperWeak));
        assertEquals(3, nonNullCount(bumperSoft));
        assertEquals(0, nonNullCount(bumperStrong));
    }

    @Test
    public void gcRedBumpers2() {
        // Given
        CarGcComponent DI = Stone.createComponent(CarGcComponent.class);
        List<WeakReference<Bumper>> bumperFactory = ListUtils.format(DI.bumpersModule().bumperFactory(), WeakReference::new);
        List<WeakReference<Bumper>> bumperWeak = ListUtils.format(DI.bumpersModule().bumperWeak(), WeakReference::new);
        List<WeakReference<Bumper>> bumperSoft = ListUtils.format(DI.bumpersModule().bumperSoft(), WeakReference::new);
        List<WeakReference<Bumper>> bumperStrong = ListUtils.format(DI.bumpersModule().bumperStrong(), WeakReference::new);

        //When
        DI.gcRedBumpers2();

        // Then
        assertEquals(0, nonNullCount(bumperFactory));
        assertEquals(0, nonNullCount(bumperWeak));
        assertEquals(3, nonNullCount(bumperSoft));
        assertEquals(0, nonNullCount(bumperStrong));
    }

    @Test
    public void gcWheelsTest() {
        // Given
        CarGcComponent DI = Stone.createComponent(CarGcComponent.class);
        List<WeakReference<Bumper>> bumperFactory = ListUtils.format(DI.bumpersModule().bumperFactory(), WeakReference::new);
        List<WeakReference<Bumper>> bumperWeak = ListUtils.format(DI.bumpersModule().bumperWeak(), WeakReference::new);
        List<WeakReference<Bumper>> bumperSoft = ListUtils.format(DI.bumpersModule().bumperSoft(), WeakReference::new);
        List<WeakReference<Bumper>> bumperStrong = ListUtils.format(DI.bumpersModule().bumperStrong(), WeakReference::new);


        //When
        DI.gcWheels();

        // Then
        assertEquals(0, nonNullCount(bumperFactory));
        assertEquals(0, nonNullCount(bumperWeak));
        assertEquals(3, nonNullCount(bumperSoft));
        assertEquals(3, nonNullCount(bumperStrong));
    }

    @Test
    public void gcNothing() {
        // Given
        CarGcComponent DI = Stone.createComponent(CarGcComponent.class);
        List<WeakReference<Bumper>> bumperFactory = ListUtils.format(DI.bumpersModule().bumperFactory(), WeakReference::new);
        List<WeakReference<Bumper>> bumperWeak = ListUtils.format(DI.bumpersModule().bumperWeak(), WeakReference::new);
        List<WeakReference<Bumper>> bumperSoft = ListUtils.format(DI.bumpersModule().bumperSoft(), WeakReference::new);
        List<WeakReference<Bumper>> bumperStrong = ListUtils.format(DI.bumpersModule().bumperStrong(), WeakReference::new);

        //When
        DI.gcNothing();

        // Then
        assertEquals(0, nonNullCount(bumperFactory));
        assertEquals(0, nonNullCount(bumperWeak));
        assertEquals(3, nonNullCount(bumperSoft));
        assertEquals(3, nonNullCount(bumperStrong));
    }


    private int nonNullCount(List<WeakReference<Bumper>> list) {
        int count = 0;
        for (int i = 0; i < list.size(); i++) {
            if (list.get(i) != null && list.get(i).get() != null) count++;
        }
        return count;
    }


}

File: tests_wraps_kotlin/src/test/kotlin/com/github/klee0kai/stone/kotlin/test/car/cachecontrol/gc/BumperGcTests.kt
package com.github.klee0kai.stone.kotlin.test.car.cachecontrol.gc

import com.github.klee0kai.stone.Stone
import com.github.klee0kai.stone.kotlin.test.utils.KotlinUtils.resetKotlinRegisters
import com.github.klee0kai.test.car.di.cachecontrol.gc.CarGcComponent
import com.github.klee0kai.test.car.model.Bumper
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import java.lang.ref.WeakReference

class BumperGcTests {
    @Test
    fun createWorkCorrect() {
        // Given
        val DI = Stone.createComponent(CarGcComponent::class.java)

        //When
        val bumperFactory = DI.bumpersModule()!!.bumperFactory().map { WeakReference(it) }
        val bumperWeak = DI.bumpersModule()!!.bumperWeak().map { WeakReference(it) }
        val bumperSoft = DI.bumpersModule()!!.bumperSoft().map { WeakReference(it) }
        val bumperStrong = DI.bumpersModule()!!.bumperStrong().map { WeakReference(it) }

        //Then
        assertEquals(3, nonNullCount(bumperFactory))
        assertEquals(3, nonNullCount(bumperWeak))
        assertEquals(3, nonNullCount(bumperSoft))
        assertEquals(3, nonNullCount(bumperStrong))
    }

    @Test
    fun gcAllTest() {
        // Given
        val DI = Stone.createComponent(CarGcComponent::class.java)
        val bumperFactory = DI.bumpersModule()!!.bumperFactory().map { WeakReference(it) }
        val bumperWeak = DI.bumpersModule()!!.bumperWeak().map { WeakReference(it) }
        val bumperSoft = DI.bumpersModule()!!.bumperSoft().map { WeakReference(it) }
        val bumperStrong = DI.bumpersModule()!!.bumperStrong().map { WeakReference(it) }

        //When
        resetKotlinRegisters()
        DI.gcAll()

        // Then
        assertEquals(0, nonNullCount(bumperFactory))
        assertEquals(0, nonNullCount(bumperWeak))
        assertEquals(0, nonNullCount(bumperSoft))
        assertEquals(0, nonNullCount(bumperStrong))
    }

    @Test
    fun gcWeakTest() {
        // Given
        val DI = Stone.createComponent(CarGcComponent::class.java)
        val bumperFactory = DI.bumpersModule()!!.bumperFactory().map { WeakReference(it) }
        val bumperWeak = DI.bumpersModule()!!.bumperWeak().map { WeakReference(it) }
        val bumperSoft = DI.bumpersModule()!!.bumperSoft().map { WeakReference(it) }
        val bumperStrong = DI.bumpersModule()!!.bumperStrong().map { WeakReference(it) }

        //When
        resetKotlinRegisters()
        DI.gcWeak()

        // Then
        assertEquals(0, nonNullCount(bumperFactory))
        assertEquals(0, nonNullCount(bumperWeak))
        assertEquals(3, nonNullCount(bumperSoft))
        assertEquals(3, nonNullCount(bumperStrong))
    }

    @Test
    fun gcSoftTest() {
        // Given
        val DI = Stone.createComponent(CarGcComponent::class.java)
        val bumperFactory = DI.bumpersModule()!!.bumperFactory().map { WeakReference(it) }
        val bumperWeak = DI.bumpersModule()!!.bumperWeak().map { WeakReference(it) }
        val bumperSoft = DI.bumpersModule()!!.bumperSoft().map { WeakReference(it) }
        val bumperStrong = DI.bumpersModule()!!.bumperStrong().map { WeakReference(it) }

        //When
        resetKotlinRegisters()
        DI.gcSoft()

        // Then
        assertEquals(0, nonNullCount(bumperFactory))
        assertEquals(0, nonNullCount(bumperWeak))
        assertEquals(0, nonNullCount(bumperSoft))
        assertEquals(3, nonNullCount(bumperStrong))
    }

    @Test
    fun gcStrongTest() {
        // Given
        val DI = Stone.createComponent(CarGcComponent::class.java)
        val bumperFactory = DI.bumpersModule()!!.bumperFactory().map { WeakReference(it) }
        val bumperWeak = DI.bumpersModule()!!.bumperWeak().map { WeakReference(it) }
        val bumperSoft = DI.bumpersModule()!!.bumperSoft().map { WeakReference(it) }
        val bumperStrong = DI.bumpersModule()!!.bumperStrong().map { WeakReference(it) }

        //When
        resetKotlinRegisters()
        DI.gcStrong()

        // Then
        assertEquals(0, nonNullCount(bumperFactory))
        assertEquals(0, nonNullCount(bumperWeak))
        assertEquals(3, nonNullCount(bumperSoft))
        assertEquals(0, nonNullCount(bumperStrong))
    }

    @Test
    fun gcBumpers() {
        // Given
        val DI = Stone.createComponent(CarGcComponent::class.java)
        val bumperFactory = DI.bumpersModule()!!.bumperFactory().map { WeakReference(it) }
        val bumperWeak = DI.bumpersModule()!!.bumperWeak().map { WeakReference(it) }
        val bumperSoft = DI.bumpersModule()!!.bumperSoft().map { WeakReference(it) }
        val bumperStrong = DI.bumpersModule()!!.bumperStrong().map { WeakReference(it) }

        //When
        resetKotlinRegisters()
        DI.gcBumpers()

        // Then
        assertEquals(0, nonNullCount(bumperFactory))
        assertEquals(0, nonNullCount(bumperWeak))
        assertEquals(0, nonNullCount(bumperSoft))
        assertEquals(0, nonNullCount(bumperStrong))
    }

    @Test
    fun gcRedBumpers() {
        // Given
        val DI = Stone.createComponent(CarGcComponent::class.java)
        val bumperFactory = DI.bumpersModule()!!.bumperFactory().map { WeakReference(it) }
        val bumperWeak = DI.bumpersModule()!!.bumperWeak().map { WeakReference(it) }
        val bumperSoft = DI.bumpersModule()!!.bumperSoft().map { WeakReference(it) }
        val bumperStrong = DI.bumpersModule()!!.bumperStrong().map { WeakReference(it) }

        //When
        resetKotlinRegisters()
        DI.gcRedBumpers()

        // Then
        assertEquals(0, nonNullCount(bumperFactory))
        assertEquals(0, nonNullCount(bumperWeak))
        assertEquals(3, nonNullCount(bumperSoft))
        assertEquals(0, nonNullCount(bumperStrong))
    }

    @Test
    fun gcRedBumpers2() {
        // Given
        val DI = Stone.createComponent(CarGcComponent::class.java)
        val bumperFactory = DI.bumpersModule()!!.bumperFactory().map { WeakReference(it) }
        val bumperWeak = DI.bumpersModule()!!.bumperWeak().map { WeakReference(it) }
        val bumperSoft = DI.bumpersModule()!!.bumperSoft().map { WeakReference(it) }
        val bumperStrong = DI.bumpersModule()!!.bumperStrong().map { WeakReference(it) }

        //When
        resetKotlinRegisters()
        DI.gcRedBumpers2()

        // Then
        assertEquals(0, nonNullCount(bumperFactory))
        assertEquals(0, nonNullCount(bumperWeak))
        assertEquals(3, nonNullCount(bumperSoft))
        assertEquals(0, nonNullCount(bumperStrong))
    }

    @Test
    fun gcWheelsTest() {
        // Given
        val DI = Stone.createComponent(CarGcComponent::class.java)
        val bumperFactory = DI.bumpersModule()!!.bumperFactory().map { WeakReference(it) }
        val bumperWeak = DI.bumpersModule()!!.bumperWeak().map { WeakReference(it) }
        val bumperSoft = DI.bumpersModule()!!.bumperSoft().map { WeakReference(it) }
        val bumperStrong = DI.bumpersModule()!!.bumperStrong().map { WeakReference(it) }


        //When
        resetKotlinRegisters()
        DI.gcWheels()

        // Then
        assertEquals(0, nonNullCount(bumperFactory))
        assertEquals(0, nonNullCount(bumperWeak))
        assertEquals(3, nonNullCount(bumperSoft))
        assertEquals(3, nonNullCount(bumperStrong))
    }

    @Test
    fun gcNothing() {
        // Given
        val DI = Stone.createComponent(CarGcComponent::class.java)
        val bumperFactory = DI.bumpersModule()!!.bumperFactory().map { WeakReference(it) }
        val bumperWeak = DI.bumpersModule()!!.bumperWeak().map { WeakReference(it) }
        val bumperSoft = DI.bumpersModule()!!.bumperSoft().map { WeakReference(it) }
        val bumperStrong = DI.bumpersModule()!!.bumperStrong().map { WeakReference(it) }

        //When
        resetKotlinRegisters()
        DI.gcNothing()

        // Then
        assertEquals(0, nonNullCount(bumperFactory))
        assertEquals(0, nonNullCount(bumperWeak))
        assertEquals(3, nonNullCount(bumperSoft))
        assertEquals(3, nonNullCount(bumperStrong))
    }

    private fun nonNullCount(list: List<WeakReference<Bumper?>?>): Int = list.count { it?.get() != null }

}
